// @logto/browser/lib/storage.js
const keyPrefix = `logto`;
class BrowserStorage {
    constructor(appId) {
        this.appId = appId;
    }
    getKey(item) {
        if (item === undefined) {
            return `${keyPrefix}:${this.appId}`;
        }
        return `${keyPrefix}:${this.appId}:${item}`;
    }
    async getItem(key) {
        if (key === 'signInSession') {
            // The latter `getItem()` is for backward compatibility. Can be removed when major bump.
            return sessionStorage.getItem(this.getKey(key)) ?? sessionStorage.getItem(this.getKey());
        }
        return localStorage.getItem(this.getKey(key));
    }
    async setItem(key, value) {
        if (key === 'signInSession') {
            sessionStorage.setItem(this.getKey(key), value);
            return;
        }
        localStorage.setItem(this.getKey(key), value);
    }
    async removeItem(key) {
        if (key === 'signInSession') {
            sessionStorage.removeItem(this.getKey(key));
            return;
        }
        localStorage.removeItem(this.getKey(key));
    }
}

export { BrowserStorage };
